### PR TITLE
Support custom domain on SSL certificate and App Registration redirect

### LINKED ISSUE #3955: Custom domain configuration?
What configuration is required in order to apply a custom domain name to a TRE instance?

I'm assuming you will need to create a DNS record to your TRE instance along the lines of:

`treinstance.mydomainname.co.uk CNAME treinstance.region.cloudapp.azure.com`

What is then required in terms of Azure TRE configuration?  Is this just setting the TRE_URL in config.yaml / env vars and redeploying?  Will the letsencrypt make action still work for custom domains?

Thanks

### PR DESCRIPTION
**Pull request by jonnyry**
# Resolves #3955

## What is being addressed

Ability to:
- deploy the TRE with a custom domain which can be used to connect to the TRE portal
- use the inbuilt Lets Encrypt job to renew the SSL certification for the domain

## How is this addressed

1. Add a new parameter `CUSTOM_DOMAIN` in config.yaml (by default commented out).

2. Feed the new parameter into the `make auth` target which adds SSO redirect URLs for the custom domain *in addition* to the default Azure one here:

https://github.com/microsoft/AzureTRE/blob/1ffb09baf37f4599adfd65b4259fdda7564da408/devops/scripts/aad/create_api_application.sh#L241-L246

3. Feed the new parameter into the `make letsencrypt` target which when populated, generates a uploads a certificate for the custom domain *instead of* the cloudapp.azure.com domain.  If not populated, it runs as the behaves as it does now generating a certificate for the default Azure domain.

4. Modify Lets Encrypt GitHub Action to look for a CUSTOM_DOMAIN GH secret and pass in to the `make letsencrypt` flow.  Again if CUSTOM_DOMAIN is present a cert is generated for it, if not it reverts to generating a cert for the default domain.

### CODE REVIEW

**Comment by marrobi** [comment]
@jonnyry  So then you are changing the domain to point to the IP off the application gateway?

Would be good if you can add a section to the docs somewhere. 

Will will get it tested. Thanks.

File: config.sample.yaml [hunk 1]
```diff
@@ -83,3 +82,7 @@ developer_settings:
 # Used by the API and Resource processor application to change log level
 # Can be "ERROR", "WARNING", "INFO", "DEBUG"
 #  logging_level: "INFO"
+
+# If you want to use TRE_URL to point to your local TRE API instance or be configured to another cloud provider
+# uncomment and set this variable
+#  tre_url: __CHANGE_ME__
```
Comment (marrobi) [inline config.sample.yaml:88]: Hmm, so this used to be used to register bundles before the TRE CLI existed.

I think it is still used in the e2e tests - see `e2e_tests/config.py`, to point to the TRE API you want to run them against.

**Comment by jonnyry** [comment]
> @jonnyry So then you are changing the domain to point to the IP off the application gateway?

Yes that's right, though I used a CNAME rather than direct to the IP:

`mytredomain.com   CNAME   mytre.region.cloudapp.azure.com`

> Would be good if you can add a section to the docs somewhere.

Sure happy to add a page in.  @tim-allen-ck mentioned a page on custom domains that was being added, I can append to that, or add a new one if not  https://github.com/microsoft/AzureTRE/issues/3955#issuecomment-2160070562

**Comment by tim-allen-ck** [comment]
> Sure happy to add a page in. @tim-allen-ck mentioned a page on custom domains that was being added, I can append to that, or add a new one if not [#3955 (comment)](https://github.com/microsoft/AzureTRE/issues/3955#issuecomment-2160070562)

I haven't made a pull request yet, so by all means go ahead and create something new, ill add those bit later

**Comment by jonnyry** [comment]
> > Sure happy to add a page in. @tim-allen-ck mentioned a page on custom domains that was being added, I can append to that, or add a new one if not [#3955 (comment)](https://github.com/microsoft/AzureTRE/issues/3955#issuecomment-2160070562)
> 
> I haven't made a pull request yet, so by all means go ahead and create something new, ill add those bit later

OK, have added in commit:  https://github.com/microsoft/AzureTRE/pull/4014/commits/5aea965dff0a835e1f61007bbf155cd44729020f

File: config.sample.yaml (hunk 1 above)
Comment (marrobi) [reply config.sample.yaml:88]: Ah, sorry, github is hiding part of the file. I see now, thats fine.

**Review by marrobi — APPROVED** [review]
LGTM

Haven't tested but from a good to have in the project am happy.

Thanks!

**Review by tim-allen-ck — APPROVED** [review]
LGTM 
Happy to merge if you've tested @jonnyry

**Comment by jonnyry** [comment]
> LGTM Happy to merge if you've tested @jonnyry

I have - running using the changes here.  Can do a quick demo if you like

**Comment by tim-allen-ck** [comment]
> > LGTM Happy to merge if you've tested @jonnyry
> 
> I have - running using the changes here. Can do a quick demo if you like

that would be great actually

**Comment by jonnyry** [comment]
> > > LGTM Happy to merge if you've tested @jonnyry
> > 
> > 
> > I have - running using the changes here. Can do a quick demo if you like
> 
> that would be great actually

OK drop an email at jonny.rylands1 [at] nhs.net

File: core/terraform/scripts/letsencrypt.sh [hunk 2]
```diff
@@ -92,6 +92,11 @@ ledir=$(pwd)/letsencrypt
 
 mkdir -p "${ledir}/logs"
 
+CERT_FQDN=$FQDN
+if [[ -n "$CUSTOM_DOMAIN" ]]; then
+  CERT_FQDN=$CUSTOM_DOMAIN
+fi
+
```
Comment (tim-allen-ck) [inline core/terraform/scripts/letsencrypt.sh:99]: could you add that log line about the domain being used?

Comment (jonnyry) [reply core/terraform/scripts/letsencrypt.sh:99]: Added

**Pull request merged**